client/app/types/course/assessment/skills/skills.ts

```diff
@@ -1,29 +1,96 @@
+export interface BranchOptions {
+  value: number;
+  label: string;
+}
+
+/**
+ * Data types for skills data retrieved from backend through API call.
+ */
+
+export interface SkillPermissions {
+  canCreateSkill: boolean;
+  canCreateSkillBranch: boolean;
+}
+
 export interface SkillData {
   id: number;
+  branchId?: number;
   title: string;
+  description?: string;
+  canUpdate?: boolean;
+  canDestroy?: boolean;
+}
+
+export interface UserSkillData extends SkillData {
   percentage: number;
   grade: number;
   totalGrade: number;
 }
 
+export interface SkillBranchData {
+  id: number;
+  title: string;
+  description?: string;
+  canUpdate?: boolean;
+  canDestroy?: boolean;
+  skills?: SkillData[];
+}
+
+export interface UserSkillBranchData extends SkillBranchData {
+  userSkills?: UserSkillData[];
+}
+
+/**
+ * Data types for achievement data used in frontend that are converted from
+ * received backend data.
+ */
+
 export interface SkillEntity {
   id: number;
+  branchId?: number;
   title: string;
+  description?: string;
+  canUpdate?: boolean;
+  canDestroy?: boolean;
+}
+
+export interface UserSkillEntity extends SkillEntity {
   percentage: number;
   grade: number;
   totalGrade: number;
 }
 
-export interface SkillBranchData {
+export interface SkillBranchEntity {
   id: number;
   title: string;
   description?: string;
-  skills?: SkillData[];
+  canUpdate?: boolean;
+  canDestroy?: boolean;
 }
 
-export interface SkillBranchEntity {
-  id: number;
+export interface UserSkillBranchEntity extends SkillBranchEntity {
+  userSkills?: UserSkillEntity[];
+}
+
+/**
+ * Data types for skills form data.
+ */
+
+export interface SkillBranchFormData {
   title: string;
-  description?: string;
-  skills?: SkillEntity[];
+  description: string;
+}
+
+export interface SkillFormData extends SkillBranchFormData {
```

Comment:
Not SkillBranchEditFormData?

Reply:
I think it is for create form data also, so should be for both?

Reply:
I think it is for create form data also, so should be for both?

Reply:
Ah i see what you are doing here.. Lets separate the type for both forms.. dont extend it because skillBranchForm != skillForm

Reply:
You also probably need to define other types: SkillBranchEditFormData and SkillEditFormData to include the id for skill branch and skill respectively, right?